text/0000-integer-overflow.md

```diff
@@ -0,0 +1,524 @@
+checks `on` or `off` solely based on performance considerations. It should be
+possible to distinguish cases where checking was too expensive from where
+wraparound was desired. (Wraparound is not usually desired.)
+
+## Scoped attributes to control runtime checking
+
+This depends on [RFC 40][40] being implemented.
+
+Introduce an `overflow_checks` attribute which can be used to turn runtime
+overflow checks on or off in a given scope. `#[overflow_checks(on)]` turns them
+on, `#[overflow_checks(off)]` turns them off. The attribute can be applied to a
+whole `crate`, a `mod`ule, an `fn`, or (as per [RFC 40][40]) a given block or
+a single expression. When applied to a block, this is analogous to the
+`checked { }` blocks of C#. As with lint attributes, an `overflow_checks`
+attribute on an inner scope or item will override the effects of any
+`overflow_checks` attributes on outer scopes or items. Overflow checks can, in
+fact, be thought of as a kind of run-time lint. Where overflow checks are in
+effect, overflow with the basic arithmetic operations and casts on the built-in
+fixed-size integer types will invoke task failure. Where they are not, the
+checks  are omitted, and the result of the operations is left unspecified (but
+will most likely wrap).
+
+Significantly, turning `overflow_checks` on or off should only produce an
+observable difference in the behavior of the program, beyond the time it takes
+to execute, if the program has an overflow bug.
+
+It should also be emphasized that `overflow_checks(off)` only disables *runtime*
+overflow checks. Compile-time analysis can and should still be performed where
+possible. Perhaps the name could be chosen to make this more obvious, such as
+`runtime_overflow_checks`, but that starts to get overly verbose.
+
+Illustration of use:
+
+    // checks are on for this crate
+    #![overflow_checks(on)]
+
+    // but they are off for this module
+    #[overflow_checks(off)]
+    mod some_stuff {
+
+        // but they are on for this function
+        #[overflow_checks(on)]
+        fn do_thing() {
+            ...
+
+            // but they are off for this block
+            #[overflow_checks(off)] {
+                ...
+                // but they are on for this expression
+                let n = #[overflow_checks(on)] (a * b + c);
+                ...
+            }
+
+            ...
+        }
+
+        ...
+    }
+
+    ...
+
+[40]: https://github.com/rust-lang/rfcs/blob/master/active/0040-more-attributes.md
+
+## Different operators
+
+Have the usual arithmetic operators check for overflow, and introduce a new set
+of operators with wraparound semantics, as done by Swift. Alternately, do the
+reverse: make the normal operators wrap around, and introduce new ones which
+check.
+
+Reasons this was not pursued: New, strange operators would pose an entrance
+barrier to the language. The use cases for wraparound semantics are not common
```

Comment:
I fully admit to having only a superficial acquaintance with systems programming, or with use cases for wrapping arithmetic in particular, which is why I chose to ask other people about it. My attitude, way back last summer when I started the mailing list thread, was, "the only use case for wrapping arithmetic I can think of off the top of my head is for hashing, but *surely* there must be many more?". That was followed by a really long discussion with a wide range of participants, in which the only other use cases anyone managed to identify were checksums and emulation. This frankly was a surprise to me, and informed the RFC. Half a year later, someone did eventually bring up counters and deltas in the other PR (which albeit still doesn't give a good idea of prevalence). Apparently "an understanding of this niche" is not so easy to come by.

Either way, this doesn't make me change my belief that "I expect this to use wraparound semantics on overflow" should (somehow) be distinguished from "I expect this not to overflow" in the source. If the former turns out to be more commonly desired than I had assumed, then more convenient syntax for it can still be added backwards compatibly at any point.

> I think it's a bit extraordinary to make the claim that they were entirely misguided by adding them

Be kind to point out where I made that claim.

I think it's a judgment call and possibly a question of different priorities. Rust has very studiously avoided introducing new symbolic operators on top of the ones C/C++ have (apart from the old built-in smart pointers). Swift, by contrast, allows user-defined operators.